meetings/working-groups/collection-literals/meetings/working-groups/collection-literals/CL-2023-06-16.md

```diff
@@ -0,0 +1,125 @@
+
+While the compiler is free to use stackallocs for these spans, it is not required to do so.  However, as some consumers *may* only want to allow collection literals when they will be stackalloc'ed we will ensure that the compiler emits a hidden-diagnostic if it heap allocs.  These consumers can set that diagnostic to be an error, blocking unwanted span heap allocations.  In those cases, they can manually stackalloc.
+
+Users who want to be able to return those spans can do so by explicit conversion to an array.  e.g. `Span<int> s = (int[])[1, 2, 3];`. 
+
+### BCL Apis
+
+BCL gave a status update on the work they have done.  Specifically: https://github.com/dotnet/runtime/pull/87945
+
+This (merged) PR adds the pattern of an immutable collection type `C<T>` having a factory method `public static C<T> C.Create<T>(ReadOnlySpan<T> values);` to create the collection.  A soon to be added attribute (`CollectionBuilderAttribute`) will be added that will allow the language/compiler to find this factory method from a given collection type.
+
+We do not plan any more BCL apis beyond this.  However, there are two key changes we will make in the compiler in terms of emit, to get even better codegen than what those above methods allow.  Specifically:
+
+1. while there is an `public static ImmutableArray<T> ImmutableArray.Create<T>(ReadOnlySpan<T> values);` method, the compiler will not call it.  Instead, it will sidestep it and use `ImmutableArray<T> ImmutableCollectionsMarshal.ToImmutableArray<T>(T[] values);` method.  This method passes ownership of the array *directly* to `ImmutableArray<T>` ensuring no overhead (no excess allocations *or* copying).
+
+2. while `List<T>` will already work with collection literals (due to it supporting collection-initializers) the compiler will emit new collection literal `List<T>'s` as:
+
+```c#
+List<int> list = [a, b, c]; // translates to:
+
+List<int> list = new List(capacity: 3);
+CollectionsMarshal.SetCount(list, 3);
+Span<int> __storage = CollectionsMarshal.AsSpan(list);
+__storage[0] = a;
+__storage[1] = b;
+__storage[2] = c;
+```
+
+Rule '1' will always apply for immutable arrays.  Rule '2' can apply depending on how complex it is for the compiler.  For example, if the code were `List<int> list] = [await a, await b, await c];` the compiler might choose to just emit this equivalently to `new List<int> { await a, await b, await c }` (i.e. equivalently to a collection initializer expression).
+
+### Breaking changes
+
+We would like to update the language so that `(A)[B]` is *always* a cast of a collection literal, not an indexing expression of a trivial identifier expression.  This would apply to `(A)[B]` (trivial identifeir), or `(A.B.C)[D]` (dotted identifier).  In both of these cases, if the user wanted indexing, it would be far more idiomatic to write this out as `A[B]` or `A.B.C[D]` (the parentheses are superfluous).
```

Comment:
```suggestion
We would like to update the language so that `(A)[B]` is *always* a cast of a collection literal, not an indexing expression of a trivial identifier expression.  This would apply to `(A)[B]` (trivial identifier), or `(A.B.C)[D]` (dotted identifier).  In both of these cases, if the user wanted indexing, it would be far more idiomatic to write this out as `A[B]` or `A.B.C[D]` (the parentheses are superfluous).
```